src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -1240,5 +1240,154 @@ private static bool HasNamespaceName(NamespaceSymbol @namespace, string namespac
+        {
+            Debug.Assert((object)type != null);
+            Debug.Assert(type.SpecialType != SpecialType.System_Void);
+
+            // Find the public static CreateAsyncMethodBuilder method.
+            // PROTOTYPE(tasklike): Look on base types.
+            var members = type.GetMembers(WellKnownMemberNames.CreateAsyncMethodBuilder);
+            foreach (var member in members)
+            {
+                if (member.Kind != SymbolKind.Method)
+                {
+                    continue;
+                }
+                var method = (MethodSymbol)member;
+                if ((method.DeclaredAccessibility == Accessibility.Public) &&
+                    method.IsStatic &&
+                    (method.ParameterCount == 0) &&
+                    !method.IsGenericMethod)
+                {
+                    var returnType = method.ReturnType as NamedTypeSymbol;
+                    if ((object)returnType == null)
+                    {
+                        break;
+                    }
+                    createBuilderMethod = method;
+                    return returnType;
+                }
+            }
+            createBuilderMethod = null;
+            return null;
+        }
+
+        /// <summary>
+        /// Replace Task-like types with Task types.
+        /// </summary>
+        internal static TypeSymbol NormalizeTaskTypes(this TypeSymbol type, CSharpCompilation compilation)
+        {
+            NormalizeTaskTypesCore(compilation, ref type);
+            return type;
+        }
+
+        /// <summary>
+        /// Replace Task-like types with Task types. Returns true if there were changes.
+        /// </summary>
+        private static bool NormalizeTaskTypesCore(CSharpCompilation compilation, ref TypeSymbol type)
+        {
+            switch (type.Kind)
+            {
+                case SymbolKind.NamedType:
+                    var namedType = (NamedTypeSymbol)type;
+                    var changed = NormalizeTaskTypesCore(compilation, ref namedType);
+                    type = namedType;
+                    return changed;
+                case SymbolKind.ArrayType:
+                    // PROTOTYPE(tasklike): Use VisitType or similar to cover all cases.
+                    break;
+            }
+            return false;
+        }
+
+        private static bool NormalizeTaskTypesCore(CSharpCompilation compilation, ref NamedTypeSymbol type)
+        {
+            var constructedFrom = type.ConstructedFrom;
+            if ((object)constructedFrom == type)
+            {
+                MethodSymbol createBuilderMethod;
+                var builderType = type.GetAsyncMethodBuilderType(out createBuilderMethod);
+                if ((object)builderType == null)
+                {
+                    return false;
+                }
+                Debug.Assert(type.Arity < 2);
```

Comment:
Why this assert? It looks like GetASyncMethodBuilderType makes no test of the arity of the tasklike.